File: src/api/restore.ts
import apiClient from './client';

export interface RestoreItemApiData {
  id: number;
  restore_job_id: number;
  backup_file_id?: number;
  relative_path: string;
  destination_path: string;
  source_size: number;
  restored_size: number;
  source_sha256?: string;
  restored_sha256?: string;
  status: string;
  retry_count: number;
  error_code?: string;
  error_message?: string;
  started_at?: string;
  completed_at?: string;
  verified_at?: string;
}

export interface RestoreJobApiData {
  id: number;
  restore_id: string;
  source_client_id: number;
  source_client_identifier?: string;
  target_client_id: number;
  target_client_identifier?: string;
  recovery_point_id: number;
  source_path: string;
  target_path: string;
  status: string;
  requested_by: string;
  restore_mode?: string;
  conflict_mode?: string;
  metadata_mode?: string;
  total_files?: number;
  completed_files?: number;
  failed_files?: number;
  skipped_files?: number;
  total_bytes?: number;
  restored_bytes?: number;
  verified_bytes?: number;
  progress_percent?: number;
  started_at?: string;
  completed_at?: string;
  cancelled_at?: string;
  restore_requested_at?: string;
  first_byte_restored_at?: string;
  error_message?: string;
  rto_metrics?: {
    queue_time: string;
    startup_time: string;
    restore_duration: string;
    total_restore_time: string;
    transfer_speed_mb_s: number;
  };
  created_at: string;
}

export interface RestorePreviewData {
  recovery_point_id: number;
  source_client_id: string;
  destination_root: string;
  restore_mode: string;
  conflict_mode: string;
  total_files: number;
  logical_bytes: number;
  estimated_stored_read_bytes: number;
  actions: {
    CREATE: number;
    OVERWRITE: number;
    SKIP: number;
    CONFLICT: number;
    RENAME: number;
  };
  items: Array<{
    backup_file_id: number;
    file_name: string;
    relative_path: string;
    destination_path: string;
    size_bytes: number;
    stored_size_bytes: number;
    sha256?: string;
    predicted_action: string;
    destination_exists: boolean;
  }>;
}

export const restoreApi = {
  list: async (params?: { status?: string }) => {
    const res = await apiClient.get('/restore/jobs', { params });
    return res.data;
  },

  get: async (restoreId: string) => {
    const res = await apiClient.get(`/restore/jobs/${restoreId}`);
    return res.data;
  },

  create: async (data: {
    source_client_id: string;
    target_client_id?: string;
    recovery_point_id: number;
    source_path: string;
    target_path: string;
    restore_mode?: string;
    conflict_mode?: string;
    metadata_mode?: string;
    selected_paths?: string[];
    acknowledge_cross_client?: boolean;
  }) => {
    const res = await apiClient.post('/restore/jobs', data);
    return res.data;
  },

  preview: async (data: {
    recovery_point_id: number;
    restore_mode: string;
    destination_root: string;
    conflict_mode: string;
    selected_paths?: string[];
  }) => {
    const res = await apiClient.post('/restore/preview', data);
    return res.data;
  },

  browseFiles: async (pointId: number) => {
    const res = await apiClient.get(`/restore/recovery-points/${pointId}/files`);
    return res.data;
  },

  searchFiles: async (pointId: number, params?: { q?: string; ext?: string }) => {
    const res = await apiClient.get(`/restore/recovery-points/${pointId}/files/search`, { params });
    return res.data;
  },

  start: async (restoreId: string) => {
    const res = await apiClient.post(`/restore/jobs/${restoreId}/start`);
    return res.data;
  },

  pause: async (restoreId: string) => {
    const res = await apiClient.post(`/restore/jobs/${restoreId}/pause`);
    return res.data;
  },

  resume: async (restoreId: string) => {
    const res = await apiClient.post(`/restore/jobs/${restoreId}/resume`);
    return res.data;
  },

  cancel: async (restoreId: string) => {
    const res = await apiClient.post(`/restore/jobs/${restoreId}/cancel`);
    return res.data;
  },

  getItems: async (restoreId: string) => {
    const res = await apiClient.get(`/restore/jobs/${restoreId}/items`);
    return res.data;
  },

  getLogs: async (restoreId: string) => {
    const res = await apiClient.get(`/restore/jobs/${restoreId}/logs`);
    return res.data;
  }
};

File: server/app/schemas/restore.py
import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, ConfigDict


class VirtualFileEntry(BaseModel):
    id: Optional[int] = None
    name: str
    path: str
    type: str  # file or directory
    size: int = 0
    sha256: Optional[str] = None
    change_type: Optional[str] = None
    modified_time: Optional[str] = None
    children: Optional[List["VirtualFileEntry"]] = None


class RestorePreviewRequest(BaseModel):
    recovery_point_id: int
    restore_mode: str = "FULL_RECOVERY_POINT"
    destination_root: str = Field(..., min_length=1)
    conflict_mode: str = "OVERWRITE"
    selected_paths: Optional[List[str]] = None


class RestorePreviewResponse(BaseModel):
    recovery_point_id: int
    source_client_id: str
    destination_root: str
    restore_mode: str
    conflict_mode: str
    total_files: int
    logical_bytes: int
    estimated_stored_read_bytes: int
    actions: Dict[str, int]
    items: List[Dict[str, Any]]


class RestoreJobCreate(BaseModel):
    source_client_id: str
    target_client_id: Optional[str] = None
    recovery_point_id: int
    source_path: str = Field(..., min_length=1)
    target_path: str = Field(..., min_length=1)
    restore_mode: str = "FULL_RECOVERY_POINT"  # FILE, FOLDER, SELECTION, FULL_RECOVERY_POINT
    conflict_mode: str = "OVERWRITE"          # SKIP, OVERWRITE, RENAME, FAIL
    metadata_mode: str = "BASIC"              # NONE, BASIC, FULL
    selected_paths: Optional[List[str]] = None
    acknowledge_cross_client: bool = False


class RestoreItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    restore_job_id: int
    backup_file_id: Optional[int] = None
    relative_path: str
    destination_path: str
    source_size: int
    restored_size: int
    source_sha256: Optional[str] = None
    restored_sha256: Optional[str] = None
    status: str
    retry_count: int = 0
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    started_at: Optional[datetime.datetime] = None
    completed_at: Optional[datetime.datetime] = None
    verified_at: Optional[datetime.datetime] = None


class RestoreJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    restore_id: str
    source_client_id: int
    source_client_identifier: Optional[str] = None
    target_client_id: int
    target_client_identifier: Optional[str] = None
    recovery_point_id: int
    source_path: str
    target_path: str
    status: str
    requested_by: str

    restore_mode: str = "FULL_RECOVERY_POINT"
    conflict_mode: str = "OVERWRITE"
    metadata_mode: str = "BASIC"

    total_files: int = 0
    completed_files: int = 0
    failed_files: int = 0
    skipped_files: int = 0

    total_bytes: int = 0
    restored_bytes: int = 0
    verified_bytes: int = 0
    progress_percent: float = 0.0

    started_at: Optional[datetime.datetime] = None
    completed_at: Optional[datetime.datetime] = None
    cancelled_at: Optional[datetime.datetime] = None
    restore_requested_at: Optional[datetime.datetime] = None
    first_byte_restored_at: Optional[datetime.datetime] = None
    error_message: Optional[str] = None

    rto_metrics: Optional[Dict[str, Any]] = None
    created_at: datetime.datetime
